src/cell_list.h

```diff
@@ -0,0 +1,230 @@
+        }
+        static constexpr int num_cells_x() {
+            return divide_round_up( MAXX, CELL_SIZE );
+        }
+        static constexpr int num_cells_y() {
+            return divide_round_up( MAXY, CELL_SIZE );
+        }
+        friend class cell_list_range< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        std::map<tripoint, T> cells[num_cells_x()][num_cells_y()];
+};
+/*
+ * A cell list range handles iteration over elements in the Cell List near some origin tripoint in
+ * nearest-first order and terminates at some specified threshold.
+ * It does this by visiting one "rank" of cells at a time.
+ * First it visits the cell the origin is in, determining the distance to the nearest cell wall,
+ * and iterating over every entity closer than that distance. This guarantees that there is no
+ * closer entity than one found via this process. Once that list is exhausted, it loads one row of
+ * cells around the origin cell and repeating the above process.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range
+{
+    public:
+        cell_list_range( const cell_list< T, CELL_SIZE, MAXX, MAXY > &list, const tripoint &p,
+                         int distance ) : list( &list ), origin( p ), origin_cell( p.xy() / CELL_SIZE ),
+            max_distance( distance ), sorted_elements( distance + 1,
+                    std::vector<const std::pair<tripoint, T >*>() ) {
+            min_bounds = origin_cell * CELL_SIZE;
+            max_bounds = min_bounds + point( CELL_SIZE - 1, CELL_SIZE - 1 );
+            distance_to_nearest_cell_wall = calc_distance();
+            // Special case for first cell.
+            add( list.cells[origin_cell.x][origin_cell.y] );
+        }
+        using iterator = cell_list_iterator<T, CELL_SIZE, MAXX, MAXY>;
+        iterator begin() {
+            return iterator( this );
+        }
+        iterator end() {
+            return iterator();
+        }
+        tripoint get_origin() const {
+            return origin;
+        }
+        int get_distance() const {
+            return max_distance;
+        }
+        friend class cell_list_iterator< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        int calc_distance() const {
+            int distance = max_distance;
+            if( min_bounds.x >= 0 ) {
+                distance = std::min( distance, origin.x - min_bounds.x );
+            }
+            if( min_bounds.y >= 0 ) {
+                distance = std::min( distance, origin.y - min_bounds.y );
+            }
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - origin.y );
+            }
+            return distance;
+        }
+        void add( const std::map<tripoint, T> &cells ) {
+            for( const std::pair<tripoint, T> &candidate : cells ) {
```

Comment:
Here `candidate` is a temporary object (the address of it is added to `sorted_elements`). Iterating over a map yields `const std::pair<const tripoint, T>` (note the second const). Otherwise you would be able to iterate over a non-const map and would have write-access to the keys of it.

Or simpler: `decltype( cells )::value_type` (or the like).

Reply:
Are you saying that since the const-ness doesn't match, I'm getting a copy instead of a reference, or that no matter what I do it's a temporary and it's invalid to capture its address?

Reply:
Evidently the former. Thanks a ton, I was nowhere near the solution.